### PR TITLE
[XLA:CPU] Clamp materialized dynamic dimension sizes to their bounds

### LINKED ISSUE #44940: [XLA] Heap corruption (free(): invalid pointer) during JIT compilation when running WhereOp with float16
XLA compilation with `tf.function(jit_compile=True)` crashes with a heap corruption error when running `tf.raw_ops.Where` with float16 input on CPU.

Eager execution correctly fails with a structured TensorFlow error:

```UNIMPLEMENTED: No WhereOp available for float16/half type on CPU```

However, when the same computation is executed under XLA compilation, TensorFlow logs successful compilation:

"Compiled cluster using XLA!"

Immediately after compilation, the process crashes with:
```
free(): invalid pointer
Aborted (core dumped)
```
or
```
Segmentation fault (core dumped)
```
Here is the [gist](https://colab.research.google.com/gist/jasminetrail/4c366f5395b5726867a1b27f1a6cc50e/issue_44938.ipynb)

---

## Minimal Code to Reproduce
```python
import traceback
import tensorflow as tf

tf.random.set_seed(4310)
t33880 = tf.random.uniform([1, 8, 8, 16], dtype=tf.float16)

def model_4310(t33880):
    t33881 = tf.raw_ops.FloorDiv(x=t33880, y=t33880)
    t33886 = tf.raw_ops.Where(condition=t33881)
    t33893 = tf.experimental.numpy.inner(t33886, t33886)
    t33894 = tf.experimental.numpy.logical_and(t33893, t33886)
    return t33894

_args = (t33880,)

print("=== EAGER ===")
try:
    eager_out = model_4310(*_args)
    print("success")
except Exception:
    print(traceback.format_exc())

print("\n=== XLA ===")
try:
    xla_out = tf.function(model_4310, jit_compile=True)(*_args)
    print("success")
except Exception:
    print(traceback.format_exc())
```

---

## Error logs

```Python
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1782587390.686813  985301 port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
I0000 00:00:1782587390.687196  985301 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
I0000 00:00:1782587390.732138  985301 cpu_feature_guard.cc:227] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI AVX512_BF16 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1782587391.895930  985301 port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
I0000 00:00:1782587391.896615  985301 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
E0000 00:00:1782587392.006309  985301 cuda_platform.cc:52] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)
E0000 00:00:1782587392.092615  985301 util.cc:131] oneDNN supports DT_HALF only on platforms with AVX-512. Falling back to the default Eigen-based implementation if present.
=== EAGER ===
W0000 00:00:1782587392.095011  985301 local_rendezvous.cc:412] Local rendezvous is aborting with status: UNIMPLEMENTED: No WhereOp available for float16/half type on CPU; dying in CPU WhereOp to avoid silently creating costly copies from device.
Traceback (most recent call last):
  File "/home/test/bugs/tf/8cd0850d.py", line 18, in <module>
    eager_out = model_4310(*_args)
  File "/home/test/bugs/tf/8cd0850d.py", line 9, in model_4310
    t33886 = tf.raw_ops.Where(condition=t33881)
  File "/home/test/.venv/lib/python3.10/site-packages/tensorflow/python/util/tf_export.py", line 377, in wrapper
    return f(**kwargs)
  File "/home/test/.venv/lib/python3.10/site-packages/tensorflow/python/ops/gen_array_ops.py", line 13174, in where
    _ops.raise_from_not_ok_status(e, name)
  File "/home/test/.venv/lib/python3.10/site-packages/tensorflow/python/framework/ops.py", line 6027, in raise_from_not_ok_status
    raise core._status_to_exception(e) from None  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.UnimplementedError: {{function_node __wrapped__Where_device_/job:localhost/replica:0/task:0/device:CPU:0}} No WhereOp available for float16/half type on CPU; dying in CPU WhereOp to avoid silently creating costly copies from device. [Op:Where] name: 


=== XLA ===
I0000 00:00:1782587392.210129  985301 service.cc:153] XLA service 0x5fcdbdd4e9a0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1782587392.210168  985301 service.cc:161]   StreamExecutor [0]: Host, Default Version (Driver: 0.0.0; Runtime: 0.0.0; Toolkit: 0.0.0; DNN: 0.0.0)
I0000 00:00:1782587392.377215  985301 device_compiler.h:208] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.
(core dumped)
```

Versions:

- Tensorflow: 2.21.0
- Python: 3.10

### PR DESCRIPTION
**Pull request by kodlan**
## 📝 Summary of Changes

When the dynamic padder materializes a dynamic tensor as a `SliceToDynamic` custom-call, the inferred runtime dimension sizes become copy-loop bounds in the generated CPU kernel, completely unchecked. Nothing validates the size operand of `set-dimension-size` against the dimension's static bound: runtime shape checks are disabled on CPU (`ShapeCheckMode::kIgnore`), and size inference propagates the operand verbatim. A frontend-produced size larger than the bound therefore makes the kernel write far outside the result allocation (heap corruption or SIGSEGV), and a negative size does the same through sign extension.

This change clamps each materialized size to `[0, bound]` (an HLO `clamp` around the size operand of the created `SliceToDynamic`). For in-contract programs the clamp is a value-level no-op; out-of-contract programs now get a defined, memory-safe truncation instead of undefined behavior, consistent with the pipeline's explicit decision to skip runtime shape checks. Actually implementing runtime shape checks (erroring on invalid sizes) would be the stricter long-term answer, but the existing `ShapeCheckMode::kIgnore` TODO in the CPU pipeline notes that checks were never implemented correctly and that models depend on them being ignored, so that is left as future work.

## 🎯 Justification

Fixes #44940. XLA CPU crashes with `free(): invalid pointer` or SIGSEGV running a TF graph with `tf.raw_ops.Where` on float16: the tf2xla bridge stamps a pre-slice runtime size (up to 1024) onto a shape sliced to bound 4 via `set-dimension-size`, and the SliceToDynamic kernel then copies 1024x4 elements into a 4x4-element buffer. Reduced to a 7-line HLO module: `set-dimension-size` with an oversize size operand segfaults `run_hlo_module` on CPU at head. Eager TensorFlow rejects the original program entirely (no f16 WhereOp on CPU), so XLA is the only path that executes it and must not corrupt memory doing so.

## 🚀 Kind of Contribution

🐛 Bug Fix

The `AlgebraicSimplifier` fold that removes redundant `SliceToDynamic(PadToStatic(x))` round trips is taught to look through the new clamp (validating that the clamp is exactly `clamp(0, size, bound)`), so passthrough graphs keep folding to the original operand and no extra copies are introduced.

## 🧪 Unit Tests:

- `dynamic_padder_test.cc` `SliceToDynamicSizeIsClamped`: the size operand of the materializing `SliceToDynamic` is clamped to `[0, bound]`.
- `DynamicLoweringTestTupleInput` expectation updated for the clamp wrapper.
- `algebraic_simplifier_test.cc` `NoOpSliceToDynamicOfPadToStaticWithClamp`: the round-trip elimination still fires with the clamped size operand.

## 🧪 Execution Tests:

- `dynamic_padder_test.cc` `ExecutionTest.OversizeDynamicDimensionSizeClamped` (CPU): an oversize runtime size returns the bound-sized result and a negative size returns an empty result; both cases crashed the process before the fix. Runs on CPU, no GPUs needed.

### CODE REVIEW

File: xla/hlo/transforms/simplifiers/algebraic_simplifier.cc [hunk 1]
```diff
@@ -5812,21 +5812,30 @@ absl::Status AlgebraicSimplifierVisitor::HandleConvert(
 
 absl::Status AlgebraicSimplifierVisitor::HandleCustomCall(
     HloInstruction* custom_call) {
-  // Remove redundant slice to dynamic of pad to static
+  // Remove redundant slice to dynamic of pad to static. The dynamic padder
```
Comment (penpornk) [inline xla/hlo/transforms/simplifiers/algebraic_simplifier.cc]: Nit: Having them as separate words together in a sentence is a bit hard to parse.
```suggestion
  // Remove redundant SliceToDynamic of PadToStatic. The dynamic padder
```

File: xla/service/dynamic_padder_test.cc [hunk 2]
```diff
@@ -316,10 +337,14 @@ ENTRY main {
   //   SliceToDynamic // Root require dynamic form tensor.
```
Comment (penpornk) [inline xla/service/dynamic_padder_test.cc:354]: Nit: Please update the comment too.

File: xla/service/dynamic_padder_test.cc [hunk 3]
```diff
@@ -1108,6 +1133,36 @@ ENTRY main {
   EXPECT_EQ(result, expected);
 }
 
+TEST_F(ExecutionTest, OversizeDynamicDimensionSizeClamped) {
```
Comment (penpornk) [inline xla/service/dynamic_padder_test.cc]: Nit: Since we also test a negative size, maybe rename this to OutOfBounds instead? (Up to you. I'm also fine with the current name.)
```suggestion
TEST_F(ExecutionTest, OutOfBoundsDynamicDimensionSizeClamped) {
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the fix and sorry for the delay! I have minor comments.

File: xla/service/dynamic_padder_test.cc (hunk 2 above)
Comment (kodlan) [reply xla/service/dynamic_padder_test.cc:354]: Done
